### PR TITLE
[cmd/opampsupervisor] Receive and report effective config

### LINKED ISSUE #30622: [cmd/opampsupervisor] Get the effective config from the OpAMP extension
Send the Collector's effective config to the OpAMP Supervisor.

### PR DESCRIPTION
**Pull request by evan-bradley**
**Link to tracking Issue:** Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30622

### CODE REVIEW

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 1]
```diff
@@ -406,24 +438,60 @@ func (s *Supervisor) startOpAMP() error {
...
+	s.logger.Debug("Starting OpAMP server...")
+	if err = s.startOpAMPServer(); err != nil {
+		return err
+	}
+	s.logger.Debug("OpAMP server started.")
+
+	return nil
+}
+
+func (s *Supervisor) startOpAMPServer() error {
```
Comment (srikanthccv) [inline cmd/opampsupervisor/supervisor/supervisor.go:464]: I am a little confused. Is there a reason to run another OpAMP server? Already one server is created in `getBootstrapInfo` and this one additionally. They attempt to use the same port?

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 2]
```diff
@@ -406,24 +438,60 @@ func (s *Supervisor) startOpAMP() error {
...
+		endpoint:         fmt.Sprintf("localhost:%d", s.opampServerPort),
+		onConnectingFunc: func(_ *http.Request) {},
+		onMessageFunc: func(_ serverTypes.Connection, message *protobufs.AgentToServer) {
+			s.logger.Debug("Received message")
+			if message.AgentDescription != nil {
+				s.agentDescription = message.AgentDescription
+			}
+			if message.EffectiveConfig != nil {
+				s.logger.Debug("Setting confmap")
+				fmt.Println(string(message.EffectiveConfig.ConfigMap.ConfigMap[""].Body))
```
Comment (srikanthccv) [inline cmd/opampsupervisor/supervisor/supervisor.go:483]:
```suggestion
```

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 3]
```diff
@@ -406,24 +438,60 @@ func (s *Supervisor) startOpAMP() error {
...
+				s.effectiveConfig.Store(string(message.EffectiveConfig.ConfigMap.ConfigMap[""].Body))
```
Comment (srikanthccv) [inline cmd/opampsupervisor/supervisor/supervisor.go:484]: Can we safely assume that the entire config is reported this way? Or should we iterate on the map and merge?

File: .chloggen/supervisor-effective-config.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: cmd/opampsupervisor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Report effective config to the OpAMP server
```
Comment (BinaryFissionGames) [inline .chloggen/supervisor-effective-config.yaml:10]:
```suggestion
note: Receive and report effective config to the OpAMP server
```

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 5]
```diff
@@ -560,17 +647,22 @@ func (s *Supervisor) loadAgentEffectiveConfig() {
...
 	cfgStr, ok := s.effectiveConfig.Load().(string)
 	if !ok {
-		cfgStr = ""
+		cfgStr, ok = s.mergedConfig.Load().(string)
+		if !ok {
+			cfgStr = ""
+		}
 	}
 
+	fmt.Println("Reporting", cfgStr)
```
Comment (BinaryFissionGames) [inline cmd/opampsupervisor/supervisor/supervisor.go:664]: Should remove this or move it to a debug log.

**Review by BinaryFissionGames — COMMENTED** [review]
Tested this a bit against BindPlane, it was able to correctly receive and report the config.

Only issue I ran into is that if the OpAMP server does not offer an a config with a key of `""`, the supervisor will panic. I think that can be addressed in a completely separate PR, though.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 1 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:474]: This server is meant to communicate with running Collector processes, as where the server in `getBootstrapInfo` is a one-shot server that is only tasked with extracting the Collector's agent description. I added some comments to explain the flow here, let me know if they explain the situation.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 3 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:493]: This is a bit of an open question. I will open an issue in the spec repo to link to here and we can use that as a starting point for discussion.

Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:493]: Spec issue: https://github.com/open-telemetry/opamp-spec/issues/184

**Review by srikanthccv — APPROVED** [review]

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 6]
```diff
@@ -218,37 +234,57 @@ func (s *Supervisor) loadConfig(configFile string) error {
...
 	if err != nil {
 		return err
 	}
 
-	var cfg bytes.Buffer
+	var k = koanf.New(".")
 
+	var cfg bytes.Buffer
 	err = s.bootstrapTemplate.Execute(&cfg, map[string]any{
 		"EndpointPort":   port,
```
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:257]: Can we rename `EndpointPort` to something more descriptive? It is now confusing that we have 2 ports that we obtain randomly and pass to the bootstrap template.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 1 above)
Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/supervisor.go:474]: >This server is meant to communicate with running Collector processes, as where the server in getBootstrapInfo is a one-shot server that is only tasked with extracting the Collector's agent description.

I am also confused and comments don't fully clarify the situation. Do we ever need the 2 opamp servers simultaneously? I think we need to bootstrap initially then the regular one once bootstrapping is finished, so why can't we reuse the same opamp server listening on the same port?

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 6 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go]: We were able to get rid of this now that we use the nop components during bootstrapping. The only port left is the one used for the OpAMP connection.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 1 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:477]: >I think we need to bootstrap initially then the regular one once bootstrapping is finished, so why can't we reuse the same opamp server listening on the same port?

I implemented it this way so we could easily embed the bootstrapping logic in the settings for the bootstrapping-only server and keep the logic for the bootstrapping process fairly linear within the Supervisor code. I may have missed a way to use the same server for both bootstrapping and regular operation, I'll see if I can find a way to eliminate the bootstrapping server.

Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/supervisor.go:477]: >I'll see if I can find a way to eliminate the bootstrapping server.

Thanks. If you find that your current approach is the right one please add some more comments to explain how/why.

Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:477]: I have some local changes to use the runtime OpAMP server to do bootstrapping, and I don't feel that there's a huge difference in cleanliness/functionality. I think my personal preference that all bootstrapping logic is isolated within a single method still remains. Using the runtime server means we need to split this logic across multiple methods and promote a few variables to struct fields, which feels harder to follow to me.

@srikanthccv @tigrannajaryan Is there an angle I am missing here? Outside of requiring an extra available port, it doesn't feel particularly wasteful to create a second server only used for bootstrapping. I'll happily defer to your preferences if this approach is harder to follow, but I like the idea of encapsulating the bootstrapping implementation.

Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:477]: To add a little clarification, the bootstrapping process requires the OpAMP server to report back to the Supervisor when the Collector has sent the AgentDescription and the bootstrapping process is complete. The difference between each approach is whether the "main" OpAMP server needs to know to do this or whether it only handles communicating with a normally-running Collector. I would rather the main server not have any knowledge of bootstrapping and focus only on running Collectors.

**Comment by cforce** [comment]
> Tested this a bit against BindPlane, it was able to correctly receive and report the config.
> 
> Only issue I ran into is that if the OpAMP server does not offer an a config with a key of `""`, the supervisor will panic. I think that can be addressed in a completely separate PR, though.

I tried it as well, but it does not work with latest bindplane and latest head supervisor

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 1 above)
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:477]: I actually like the idea. I do not find the separate server for bootstrapping confusing now.

Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:479]: We discussed this at the 2024-04-30 OpAMP WG meeting and determined we're okay with the current approach.

**Comment by evan-bradley** [comment]
Closing in favor of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33462.

**Pull request closed without merging**